### PR TITLE
chore: move updateBuildMetadata into its own...

### PR DESCRIPTION
**Pull request by nickboldt**
### What does this PR do?

chore: move updateBuildMetadata into its own script so we can ccall it more easily when we want to bump the timestamp as part of a build

Signed-off-by: Nick Boldt <nboldt@redhat.com>

include versions:metadata when doing prettier:fix and build commands

Signed-off-by: Nick Boldt <nboldt@redhat.com>

### Screenshot/screencast of this PR
N/A

### What issues does this PR fix or reference?
N/A (or see commit message above for issue number)

### How to test this PR?
N/A

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [x] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [x] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [x] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [ ] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [x] [The repository devfile is up to date and works, if one exists](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [x] [Sections What issues does this PR fix or reference and How to test this PR completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [x] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [x] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [x] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### CODE REVIEW

**Review by schultzp2020 — COMMENTED** [review]
I think there is a better way of updating the metadata then including under `build` and `prettier:fix`. Maybe we can have some sort of post merge commit that runs this?

File: scripts/update-metadata.mjs [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+const __dirname = dirname(__filename);
+
+// Change directory to the root of the project
+process.chdir(join(__dirname, '..'));
+
+const backstageVersion = JSON.parse(
+  readFileSync('backstage.json', 'utf8'),
+).version.toString();
+
+function updateBuildMetadata(backstageVersion) {
```
Comment (schultzp2020) [inline scripts/update-metadata.mjs:17]:
```suggestion
export function updateBuildMetadata(backstageVersion) {
```

File: scripts/update-metadata.mjs [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+  buildMetadata.card = card;
+
+  writeFileSync(
+    buildMetadataPath,
+    JSON.stringify(buildMetadata, null, 2),
+    'utf8',
+  );
+}
+
+export { updateBuildMetadata };
```
Comment (schultzp2020) [inline scripts/update-metadata.mjs:50]:
```suggestion
```

File: .rhdh/docker/Dockerfile [hunk 3]
```diff
@@ -135,7 +135,9 @@ RUN $YARN build --filter=backend && \
   # Already imported the packages above; need to `yarn install` on the `dist-dynamic` sub-folder for backend plugins
   $YARN --cwd ./dynamic-plugins/imports install-dynamic && \
   $YARN export-dynamic -- --filter=./dynamic-plugins/wrappers/* && \
-  $YARN copy-dynamic-plugins dist
+  $YARN copy-dynamic-plugins dist && \
+  # insert build time into the build-metadata.json
+  now=$(date -u +%FT%TZ); sed -i packages/app/src/build-metadata.json -r -e "s/\"(Last Commit: )(.+)\"/\"\1\2\", \"Build Time: $now\""
```
Comment (schultzp2020) [inline .rhdh/docker/Dockerfile:140]: why don't you just run the `versions:metadata` script here?

File: docker/Dockerfile [hunk 4]
```diff
@@ -71,7 +71,9 @@ RUN $YARN build --filter=backend && \
   # Already imported the packages above; need to `yarn install` on the `dist-dynamic` sub-folder for backend plugins
   $YARN --cwd ./dynamic-plugins/imports install-dynamic && \
   $YARN export-dynamic -- --filter=./dynamic-plugins/wrappers/* && \
-  $YARN copy-dynamic-plugins dist
+  $YARN copy-dynamic-plugins dist && \
+  # insert build time into the build-metadata.json
+  now=$(date -u +%FT%TZ); sed -i packages/app/src/build-metadata.json -r -e "s/\"(Last Commit: )(.+)\"/\"\1\2\", \"Build Time: $now\""
```
Comment (schultzp2020) [inline docker/Dockerfile:76]: Same here

File: .rhdh/docker/Dockerfile (hunk 3 above)
Comment (nickboldt) [reply .rhdh/docker/Dockerfile:140]: I'm adding `Build Time:` not changing `Last Commit`. For this is the BUILD time, not the last commit time. :D

File: docker/Dockerfile (hunk 4 above)
Comment (nickboldt) [reply docker/Dockerfile:76]: if versions:metadata could optionally add a second date, then sure. we could use that instead.

**Review by schultzp2020 — APPROVED** [review]
/lgtm
/approve

File: scripts/update-metadata.mjs [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+  const buildMetadata = JSON.parse(readFileSync(buildMetadataPath, 'utf8'));
+
+  const rhdhVersion = JSON.parse(
+    readFileSync('package.json', 'utf8'),
+  ).version.toString();
+
+  // const commitTime = new Date().toISOString() is similar but returns millis too: 2024-05-03T12:12:08.174Z
+  // using the shell command allows us to have consistent datestamping everywhere, including sed+date transforms downstream
+  // unsure if this will break anyone developing on Windows; could switch to ISOString if it's a problem.
+  const commitTime = execSync('/usr/bin/date -u +%FT%TZ').toString().trim(); // eg., 2024-05-03T12:12:08Z
```
Comment (ciiay) [inline scripts/update-metadata.mjs:34]: @nickboldt I'm using macOS which doesn't have a `/usr/bin/date` file. Instead it's `/bin/date`.
```
> root@1.2.0 versions:metadata
> node ./scripts/update-metadata.mjs

Updating packages/app/src/build-metadata.json ...
/bin/sh: /usr/bin/date: No such file or directory
node:internal/errors:865
  const err = new Error(message);
              ^

Error: Command failed: /usr/bin/date -u +%FT%TZ
/bin/sh: /usr/bin/date: No such file or directory
```

Comment (subhashkhileri) [reply scripts/update-metadata.mjs:34]: Blocked!!, I am also a Mac user, and I'm getting the same error and unable to make a commit. @nickboldt

Comment (nickboldt) [reply scripts/update-metadata.mjs:34]: Workaround:

don't run this pre-commit hook:

```
$➔ cat .husky/pre-commit 
#!/usr/bin/env sh
. "$(dirname -- "$0")/_/husky.sh"

npm run lint-staged
npm run versions:metadata
```

instead run this:

```
$➔ cat .husky/pre-commit 
#!/usr/bin/env sh
. "$(dirname -- "$0")/_/husky.sh"

npm run lint-staged
# npm run versions:metadata
```

Comment (nickboldt) [reply scripts/update-metadata.mjs:34]: this should work on all OSes. 

https://github.com/janus-idp/backstage-showcase/pull/1231/files

**Pull request merged**